clang/test/SemaCXX/attr-diagnose-specializations.cpp

```diff
@@ -0,0 +1,34 @@
+// RUN: %clang_cc1 %s -verify
+
+#if !__has_cpp_attribute(clang::diagnose_specializations)
+#  error
+#endif
+
+struct [[clang::diagnose_specializations]] S {}; // expected-warning {{'diagnose_specializations' attribute only applies to class templates}}
+
+template <class T, class U>
+struct [[clang::diagnose_specializations]] is_same {
```

Comment:
Would it make sense to allow passing a string (maybe optionally) to explain why the template shouldn't be specialized? Then from libc++ we would do:

```
#define _LIBCPP_NO_SPECIALIZATION [[_Clang::__diagnose_specializations__("Per <section> of the C++ Standard, users are not allowed to specialize class templates in namespace std")]]
```

Reply:
I guess we could add an optional string, but I'd actually keep the error message generic in libc++. IMO adding the stable name is more useful for us than for users. Maybe a bit stronger wording like `'X' cannot be specialized` would make sense.